Add secure renegotiation to DTLS 1.2

julek-wolfssl opened the pull request:
- Hash of fragmented certificate was not calculated as a single message and instead we were hashing individual fragments which produced the wrong digest, shared secret, etc...
- Reset handshake number after server Finished packet is sent or received (depending on side)
- Reserve space in buffer for cipher stuff
- Take `DTLS_RECORD_EXTRA` and  `DTLS_HANDSHAKE_EXTRA` into size and offset calculations for DTLS path
- Fix renegotiation in DTLS with AES128-SHA
- Fix renegotiation in DTLS with AES-GCM
- Support HelloVerify request during secure renegotiation
- Save renegotiation handshake messages for retransmission in timeout
- Handle cipher parameters from different epochs. DTLS may need to resend and receive messages from previous epochs so handling different sets of encryption and decryption parameters is crucial.

julek-wolfssl committed:
Add secure renegotiation to DTLS 1.2

- Hash of fragmented certificate was not calculated as a single message and instead we were hashing individual fragments which produced the wrong digest, shared secret, etc...
- Reset handshake number after server Finished packet is sent or received (depending on side)
- Reserve space in buffer for cipher stuff
- Take `DTLS_RECORD_EXTRA` and  `DTLS_HANDSHAKE_EXTRA` into size and offset calculations for DTLS path
- Fix renegotiation in DTLS with AES128-SHA
- Fix renegotiation in DTLS with AES-GCM
- Support HelloVerify request during secure renegotiation
- Save renegotiation handshake messages for retransmission in timeout
- Handle cipher parameters from different epochs. DTLS may need to resend and receive messages from previous epochs so handling different sets of encryption and decryption parameters is crucial.

julek-wolfssl committed:
Comments and formatting

julek-wolfssl committed:
Jenkins tests fixes

ejohnstown reviewed (changes requested):
I'm still reviewing, but I'm giving you something to work with (or respond that is intentional).

julek-wolfssl commented on src/tls.c:
```diff
@@ -1217,9 +1226,19 @@ int TLS_hmac(WOLFSSL* ssl, byte* digest, const byte* in, word32 sz, int padSz,
+#endif
     ret = wc_HmacSetKey(&hmac, wolfSSL_GetHmacType(ssl),
-                                              wolfSSL_GetMacSecret(ssl, verify),
+                                              macSecret,
```
Why though?

ejohnstown replied:
We have been clearing pointers elsewhere when we're done with them, mostly when freeing them. This one is pointing at the master secret. I don't know. It made sense to me at the time.

julek-wolfssl replied:
It's a local variable so I'm not sure if null'ing it is that important.

julek-wolfssl committed:
Restore StoreKeys functionality for TLS case

julek-wolfssl commented on src/keys.c:
```diff
@@ -3209,11 +3222,25 @@ int StoreKeys(WOLFSSL* ssl, const byte* keyData, int side)
 #ifdef HAVE_SECURE_RENEGOTIATION
-    if (ssl->secure_renegotiation && ssl->secure_renegotiation->cache_status ==
-                                                            SCR_CACHE_NEEDED) {
+    if (ssl->options.dtls &&
```
It should retain the same functionality as before. I changed it back in a new commit.

toddouska commented on src/tls.c:
```diff
@@ -1197,7 +1206,7 @@ int TLS_hmac(WOLFSSL* ssl, byte* digest, const byte* in, word32 sz, int padSz,
 #endif
 
-    wolfSSL_SetTlsHmacInner(ssl, myInner, sz, content, verify);
+    wolfSSL_SetTlsHmacInner(ssl, myInner, sz, content, epochOrder);
```
How can epochOrder mean the same thing as verify hmac side or create hmac side?

julek-wolfssl replied:
This is an incorrect naming in the `wolfSSL_SetTlsHmacInner` function. The `verify` argument is passed directly to `WriteSEQ` as `verifyOrder` and then to `DtlsGetSEQ` as `order` which uses it to determine which epoch and sequence numbers should be used.
The name `verify` makes sense in the TLS (in `GetSEQIncrement`) case since it only determines if we want to verify for sending or receiving. This worked coincidentally for `DtlsGetSEQ` because of how the orders are defined:
```
    PREV_ORDER         = -1,   /* Sequence number is in previous epoch. */
    PEER_ORDER         = 1,    /* Peer sequence number for verify. */
    CUR_ORDER          = 0,    /* Current sequence number. */
```
So if `verify = 1` then this means receiving or `PEER_ORDER` while `verify = 0` means sending or `CUR_ORDER`. This obviously doesn't take the possibility of `PREV_ORDER` into account so I had to modify this part.

toddouska replied:
epochOrder is not the correct name.  epoch does NOT exist in TLS.  We can't rely on side effects / interactions with other items.  That is, if you're trying to combine the verify side and a DTLS order in the same variable, separate them into 2 variables.  Or have a much clearer name, explanation.  Remember, someone has to easily read / maintain / understand each of these items exactly.

ejohnstown commented:
Can you add test cases to the test-dtls.conf for secure renegotiation? Check test.conf for some examples.

toddouska commented on src/internal.c:
```diff
@@ -228,15 +228,25 @@ static WC_INLINE int IsEncryptionOn(WOLFSSL* ssl, int isSend)
-    return result;
+int IsInitialRenegotiationState(WOLFSSL* ssl)
+{
+    if (ssl->options.acceptState == ACCEPT_FIRST_REPLY_DONE
```
How does ACCEPT_FIRST_REPLY_DONE guarantee Renegotiation?

julek-wolfssl replied:
`IsInitialRenegotiationState` is used in `DoClientHello` to determine whether or not the `HelloVerifyReqeust` is necessary. I decided (with John's approval) that a `HelloVerifyReqeust` is unnecessary for a renegotiation. In `DoHandShakeMsgType` if we receive a `client_hello` and have `ssl->secure_renegotiation = 1` then `ssl->options.acceptState` is set to `ACCEPT_FIRST_REPLY_DONE`. From testing this is a sufficient signal for `DoClientHello` to know that it doesn't need to compute the Hmac.

toddouska replied:
But this function doesn't test secure_renegotation status.  We shouldn't rely on someone testing that first either.

toddouska commented on src/internal.c:
```diff
@@ -6731,6 +6741,14 @@ static WC_INLINE void GetSEQIncrement(WOLFSSL* ssl, int verify, word32 seq[2])
 static WC_INLINE void DtlsGetSEQ(WOLFSSL* ssl, int order, word32 seq[2])
 {
+#ifdef HAVE_SECURE_RENEGOTIATION
+    /* if ssl->secure_renegotiation->tmp_keys.dtls_epoch > ssl->keys.dtls_epoch then PREV_ORDER
```
This pattern is in a few places.  Is there a way to isolate the behavior to a single function?

toddouska commented on src/internal.c:
```diff
@@ -7191,12 +7249,39 @@ int VerifyForDtlsMsgPoolSend(WOLFSSL* ssl, byte type, word32 fragOffset)
 }
 
 
+int VerifyForTxDtlsMsgDelete(WOLFSSL* ssl, DtlsMsg* item)
```
Please add a comment describing what this new function does.

toddouska commented on src/internal.c:
```diff
@@ -15284,6 +15506,25 @@ int ProcessReply(WOLFSSL* ssl)
 
                     ssl->buffers.inputBuffer.idx++;
+
+                    ret = SanityCheckMsgReceived(ssl, change_cipher_hs);
```
Should this sanity check be moved down this far?

julek-wolfssl replied:
Previously the packet checking was being done after the sanity fails or succeeds but in an incorrect way (for DTLS). In the case of sanity check failure the result of the packet check would be returned (assuming `DUPLICATE_MSG_E` or `OUT_OF_ORDER_E` was returned by the sanity check). I unified the packet checking and moved the sanity check immediately after confirming that we have received a correct packet.

toddouska commented on src/internal.c:
```diff
@@ -16558,6 +16905,8 @@ int SendCertificate(WOLFSSL* ssl)
         #endif
         }
 
+        sendSz += cipherExtraData(ssl);
```
Even when encryption is off?

dgarske reviewed (changes requested):
It fails with asynchronous mode enabled.

```
./configure --enable-asynccrypt --enable-dtls --enable-secure-renegotiation
make

./tests/unit.test

trying server command line[422]: SuiteTest -v 3 -l ECDHE-RSA-AES128-GCM-SHA256 -M -2 -p 0
trying client command line[423]: SuiteTest -v 3 -l ECDHE-RSA-AES128-GCM-SHA256 -i -2 -p 33951
SSL version is TLSv1.2
SSL cipher suite is TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
SSL curve name is FFDHE_2048
SSL version is TLSv1.2
SSL cipher suite is TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
SSL curve name is SECP256R1
Beginning secure rengotiation.
err = -108, wolfCrypt Operation Pending (would block / eagain) error
wolfSSL error: wolfSSL_Rehandshake failed
```

You can use / see the `./async-check.sh` script for getting the required async files to reproduce.

julek-wolfssl commented:
DTLS handshake fails when used with `wolfSSL_CTX_set_group_messages`.

julek-wolfssl committed:
DTLS Message Grouping

Flush output buffer when we suspect that the grouped messages may exceed MTU.

julek-wolfssl commented:
Fixing the grouping took some time today because of an interesting phenomenon with UDP. When reading only part of a datagram Linux drops the rest of the datagram and the next read returns data from the next received datagram.

julek-wolfssl committed:
DTLS test cases

julek-wolfssl committed:
Fix asynchronous DTLS issue

julek-wolfssl committed:
WIP

julek-wolfssl commented:
@dgarske it looks like async is broken for secure renegotiation in master as well (for both DTLS and TLS). I'll fix it in this PR.

julek-wolfssl committed:
WIP

julek-wolfssl committed:
ASYNC: Working AES128-SHA

julek-wolfssl committed:
ASYNC: Working TLS SCR

julek-wolfssl committed:
ASYNC: Fix issues with TLS and DTLS

julek-wolfssl committed:
Fix SessionTicket length in unencrypted case

julek-wolfssl committed:
Fix Jenkins

julek-wolfssl committed:
Missing ssl->heap in FreeBuildMsgArgs

julek-wolfssl committed:
Code Review

dgarske reviewed (approved):
@julek-wolfssl : Changes look good and test out correctly. Also tested with Intel QuickAssist asynchronous mode.

julek-wolfssl committed:
Ignore duplicate or out of order CCS message

Init variables since compiler complains they might be used without initialization.
